astroquery/mast/core.py

```diff
@@ -1270,6 +1278,91 @@ def _download_curl_script(self, products, out_dir):
         return manifest
 
+    def enable_s3_hst_dataset(self):
+        """
+        Attempts to enable downloading HST public files from S3 instead of MAST.
+        Requires the boto3 library to function.
+        """
+        import boto3
+        self._boto3 = boto3
+
+        log.info("Using the S3 HST public dataset")
+        log.warning("Your AWS account will be charged for access to the S3 bucket")
+        log.info("See Request Pricing in https://aws.amazon.com/s3/pricing/ for details")
+        log.info("If you have not configured boto3, follow the instructions here: " +
+                 "https://boto3.readthedocs.io/en/latest/guide/configuration.html")
+
+    def disable_s3_hst_dataset(self):
+        """
+        Disables downloading HST public files from S3 instead of MAST
+        """
+        self._boto3 = None
+
+    def _download_from_s3(self, dataProduct, localPath, cache=True):
+        # The following is a mishmash of BaseQuery._download_file and s3 access through boto
+
+        bkt_name = 'stpubdata'
+
+        # This is a cheap operation and does not perform any actual work yet
+        s3 = self._boto3.resource('s3')
+        s3_client = self._boto3.client('s3')
+        bkt = s3.Bucket(bkt_name)
+
+        dataUri = dataProduct['dataURI']
+        obs_id = dataProduct['obs_id']
+
+        obs_id = obs_id.lower()
+
+        # magic associations logic per Brian - 0-9/a-e we convert to 0
+        magicValues = "123456789abcde"
+        if obs_id[-1] in magicValues:
+            new_obs_id = obs_id[:-1] + "0"
+            dataUri = dataUri.replace(obs_id, new_obs_id)
+            obs_id = new_obs_id
+            log.warning("This data product's path may not have been properly identified %s" % dataUri)
+
+        bucketPath = "hst/public/" + obs_id[:4] + dataUri.replace("mast:HST/product", "")
+
+        # CAOM LIES!
+        # We can't use the reported file size
+        # length = dataProduct["size"]
+        info_lookup = s3_client.head_object(Bucket=bkt_name, Key=bucketPath, RequestPayer='requester')
+        length = info_lookup["ContentLength"]
+
+        if cache and os.path.exists(localPath):
+            if length is not None:
+                statinfo = os.stat(localPath)
+                if statinfo.st_size != length:
+                    log.warning("Found cached file {0} with size {1} that is "
+                                "different from expected size {2}"
+                                .format(localPath,
+                                        statinfo.st_size,
+                                        length))
+                else:
+                    log.info("Found cached file {0} with expected size {1}."
+                             .format(localPath, statinfo.st_size))
+                    return
+
+        with ProgressBarOrSpinner(length, ('Downloading URL s3://{0}/{1} to {2} ...'.format(
+                bkt_name, bucketPath, localPath))) as pb:
+
+            global bytes_read
```

Comment:
Is this global required because of the use of threading below?  Is this use of `threading` necessary?  Could you also clarify exactly what should happen here: is this going to be an independent thread running in the background, with control returned to the main thread?  If not, why include it?

(I have not used threading much)

Reply:
The callback function is called from multiple threads created by boto3 when downloading a file.  I can put some notes in here

Reply:
Please do, thanks - I didn't realize boto was doing something with threading.  Your design now makes sense.